KEP-3715: Elastic Indexed Jobs (#3724)
kubernetes/enhancements: keps/sig-apps/3715-elastic-indexed-job/README.md

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently `spec.completions` is an immutable field in Job for both `Indexed` and `NonIndexed` modes. 
This KEP proposes to allow mutating `spec.completions` for `Indexed` job iff the value in 
the update is equal to `spec.parallelsim` specified in the update request. 

Success and failure semantics are not changed for jobs that never mutate `spec.completions`; 
however, for ones that do, the semantics are changed in the following manner:

Failures will be counted against the lifetime of the job. If completions was scaled down, then previous pod failures
outside the new range will still count against the job's backoffLimit. Morever, the field `status.Failed` will not be
decremented if previously failed pods are now outside the range. However, `status.Succeeded` will be updated to reflect
the number of successful indexes within the range as defined below.

In the case where an index successfully completes, and then the job is scaled down rendering the index
out of range, and then the job is scaled up again including back the index in the range, then the index 

> will be restarted.

Reviewer: is this the current behavior of the controller in 1.26?

Reviewer 2: > Currently `spec.completions` is an immutable field in Job for both `Indexed` and `NonIndexed` modes. 
This KEP proposes to allow mutating `spec.completions` for `Indexed` job iff the value in 
the update is equal to `spec.parallelsim` specified in the update request.

From this statement, I understand if parallelism is `P` and completions are `C`, the only new value that can be given for completions is `P`. If this is correct, I have a doubt about the following statement.

> In the case where an index successfully completes, and then the job is scaled down rendering the index
out of range, and then the job is scaled up again including back the index in the range, then the index 
will be restarted.

How can the same job be scaled down and scaled up? Initially assume that the job has parallelism `P` and completions `P+K`. On scaling down the completions, the completion becomes `P`. How do we scale up again? Wouldn't scaling up again fail since the new value of completions will be different from `P`?

Am I missing something?

> 

Author: @soltysh @liggitt are we on agreement on the current semantics? The summary captures the latest proposal.

Reviewer: lgtm

Reviewer 3: lgtm as well

Author: Thanks Maciej and Jordan!

seems to not be an issue because the job controller uses Update when setting the
finished status on the job object, which is subject to object modified errors if the spec
changes at the same time.

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

The controller already counts the number of completed indexes within the range defined by `spec.completions` and
uses that to evaluate if the job completed successfully and to set the `status.Succeeded` count. This is not impacted
if `spec.completions` changes over time. Morever, on scale down, the controller also automatically removes any indexes
outside the range. In job stat

As mentioned above, pod failures may have happened to indexes that are outside the range after `spec.completions`
is updated. Those failures will still count against the backoffLimit of the job and the field `status.Failed` will
not be decremented.

> 

Author: @liggitt @soltysh since the change is only relaxing validation, I suggest we start the feature in beta and enabled by default right away. Just like we successfully did for https://github.com/kubernetes/enhancements/issues/2926.

Any objections?

Author: > The next question to ask is what a 1.26-level controller would do with a job that mutated its completions/parallelism.

It should handle it correctly. In fact, as of writing this KEP, 1.27 and 1.26 are identical.

If tests reveal that we need to make changes rendering the above not valid, then we will limit it to alpha.

Reviewer: > It should handle it correctly. In fact, as of writing this KEP, 1.27 and 1.26 are identical.

that's great to hear

> If tests reveal that we need to make changes rendering the above not valid, then we will limit it to alpha.

sgtm

Reviewer 4: Do we already have a log that trims the `CompletedIndexes` field appropriately? Are we just computing it from scratch every time?

Author: It gets re-calculated on every sync and stored in status if it changes.

Reviewer 5: If by computing it from scratch you mean that we look at all the existing pods, the answer is NO. We fixed that with tracking with finalizers :smiley: 

But we do parse the field, and update accordingly, considering the newly finished pods.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently `spec.completions` is an immutable field in Job for both `Indexed` and `NonIndexed` modes. 
This KEP proposes to allow mutating `spec.completions` for `Indexed` job iff the value in 

> the update is equal to `spec.parallelsim` specified in the update request. 

Author: not necessarily before

> the update is equal to `spec.parallelsim` specified in the update request. 

Reviewer 5: I know the controller could handle it, but it sounds a bit arbitrary that you can get from parallelism!=completions to parallelism==completions, but not back.

Semantically, it makes more sense to me that they are only mutable if the fields are matching from the beginning.

Success and failure semantics are not changed for jobs that never mutate `spec.completions`; 
however, for ones that do, the semantics are changed in the following manner:

> Failures will be counted against the lifetime of the job. If completions was scaled down, then previous pod failures

Author: done.

outside the new range will still count against the job's backoffLimit. Morever, the field `status.Failed` will not be
decremented if previously failed pods are now outside the range. However, `status.Succeeded` will be updated to reflect
the number of successful indexes within the range as defined below.

> In the case where an index successfully completes, and then the job is scaled down rendering the index

Reviewer 2: > Currently `spec.completions` is an immutable field in Job for both `Indexed` and `NonIndexed` modes. 
This KEP proposes to allow mutating `spec.completions` for `Indexed` job iff the value in 
the update is equal to `spec.parallelsim` specified in the update request.

From this statement, I understand if parallelism is `P` and completions are `C`, the only new value that can be given for completions is `P`. If this is correct, I have a doubt about the following statement.

> In the case where an index successfully completes, and then the job is scaled down rendering the index
out of range, and then the job is scaled up again including back the index in the range, then the index 
will be restarted.

How can the same job be scaled down and scaled up? Initially assume that the job has parallelism `P` and completions `P+K`. On scaling down the completions, the completion becomes `P`. How do we scale up again? Wouldn't scaling up again fail since the new value of completions will be different from `P`?

Am I missing something?

Author: completions and parallelism can be updated together at the same time, up and down, and that is the typical scenario for elastic jobs.

Reviewer 2: Does this mean the parallelism and completions are mutated together with the same value?

Author: Correct.

headless service is created to set up stable hostnames for the worker pods . The workers job is
scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.

The success semantics of these workloads are tied to the manager, not the workers. However, in
such workloads, when the job is finished the workers are either deleted directly (effectively scaled down to 0)
or they potentially exit on their own after communicating their final status to the manager who decides the final status
of the whole workload.

<!--

### Notes/Constraints/Caveats (Optional)

What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

There is a concern that changing completion semantics on the fly can be subject to
race conditions with declaring the job finished. However, after a closer look, this 
seems to not be an issue because the job controller uses Update when setting the
finished status on the job object, which is subject to object modified errors if the spec
changes at the same time.

> <!--

Reviewer 5: Let's also add that `completions` will be immutable once a job has a finished condition.

What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> ## Design Details

Reviewer 5: Add a subsection to clarify the validation logic.

Author: added a paragraph at the end.

The controller already counts the number of completed indexes within the range defined by `spec.completions` and
uses that to evaluate if the job completed successfully and to set the `status.Succeeded` count. This is not impacted
if `spec.completions` changes over time. Morever, on scale down, the controller also automatically removes any indexes

> outside the range. In job stat

Reviewer 2: `In job stat` - Incomplete sentence?

implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `k8s.io/kubernetes/pkg/controller/job`: `1/10/2023` - `90%`
- `k8s.io/kubernetes/pkg/apis/batch/validation`: `1/10/2023` - `96%`

##### Integration tests

For an `Indexed` Job test that:
- `spec.completions` is only mutable if equal to `spec.parallelism`
- When scaled down, higher indexes are removed first.

> - Success, failure and completion semantics are as discussed above.

Reviewer 5: Also test scale up.

Any other challenging scenarios for the controller you can think of?

Reviewer 5: Scale down to zero.

Author: added.

  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently `spec.completions` is an immutable in Job. This KEP proposes a new
`CompletionMode` named `ElasticIndexed` that behaves similar to `Indexed` Job,

> but allows to mutate both `spec.completions` and `spec.parallelsim` if done together.

Reviewer 3: The risks for mutating `.spec.completions` still hold and based on the discussions we've had so far I don't see strong arguments towards relaxing that. How do you expect a controller to react if it's just marking a job as completed, while a user at the same time modifies completions?

Reviewer: > How do you expect a controller to react if it's just marking a job as completed, while a user at the same time modifies completions?

We're already relying on the workloads in an autoscaling job to be coherent and not complete with success until the work is actually done, so I'm not sure the "appears complete" and "needs to scale up" race would happen much in practice, but for correctness, I would expect server validation and controller preconditions to prevent the race:
* controller marking complete would get a resourceVersion conflict error on write if the user had since mutated completions/parallelism
* user mutating completions/parallelism would be rejected if the job is already complete

## Motivation

There are cases where a batch workload requires an autoscaled indexed job with stable 
DNS pod names, for example MPI Horovord, Ray, PyTorch etc. This is currently not possible 
because `spec.completions`, which controls the range of indexes in `Indexed` job, is immutable. 
While such workloads could be modeled as a StatefulSet, the Job API is a better fit because it
offers batch-specific features such as allowing indexes to run to completion, and [pod and container failure handling](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures).

In the above mentioned cases, there is no distinction between the concept of "how many completed 
indexes do I need" (i.e., `spec.completions`) and the concept of "how many pods do I want running
at a time" (i.e., `spec.parallelism`). All indexes are expected to start at the same time, but they
can finish at different times.

The above behavior can be modeled using an Indexed job with
- `spec.completions` equal to `spec.parallelism`
- mutable `spec.completions` and `spec.parallelsim` as long as they are mutated together.

> To represent this mode of operation for Job, this KEP introduces a new

Reviewer 3: From a UX point of view I don't see any value from forcing user to always set both of these fields to the exact same value, rather than leaving `.spec.completions` set to `nil` and immutable and only modifying `.spec.parallelism` (as it's currently with non-index job). 
I totally understand @liggitt's concerns about breaking users' expectations about modifying the current IndexedJobs `.spec.completions` (and as usual he's probably right :wink: ) but I'm not sure that throwing the burden of setting both fields to user is a good option here. Especially that we are introducing entirely new CompletionMode.

Author: I would also have preferred if this case is possible by modifying a single field, but I don't see a major problem of having two to be consistent with the general case. Especially if the scale subresource idea doesn't have a good chance in the future.

-->

#### Story 1

I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.

This workload can be modeled as two Jobs, an `Indexed` Job for the manager and an `ElasticIndexed` Job 
for the workers. A headless service is created to set up stable hostnames for the worker pods . The workers job is
scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.

<!--

### Notes/Constraints/Caveats (Optional)

What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

There is a concern that changing completion semantics on the fly can be subject to
race conditions with declaring the job finished. However, after a closer look, this 
seems to not be an issue because the job controller uses Update when setting the

> finished status on the job object, which is subject to object modified errors if the spec

Reviewer 3: Right, but that suggests you're basically hoping that the user is sooner with his change than controller, what if user performs `kubectl edit` and the controller happens to be faster, but since `kubectl edit` issues a patch under the covers both options will be applied, iow. the controller finishing a job and user changing completions. I'd still advise against mutating completions.

Reviewer: See above comment about relying on coherent workloads in an autoscaling indexed job that don't complete with success while the job would still need to scale

Reviewer 3: Moving to https://github.com/kubernetes/enhancements/pull/3724#discussion_r1082521252

changes at the same time.

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

> On the job controller side, the only required change is to ensure that `ElasticIndexed` is treated as an `Indexed` Job. 

Reviewer 3: If we're going with a new completion mode, why do we want to speed up the implementation, what's wrong with the usual alpha->beta->ga process?

Author: This is not accelerating, this is required to safely add a new enum

Reviewer 3: I'd consider new enum, as a new feature, thus I'm not sure I understand the need for not going the usual path?

Author: This is not an unusual path, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field

but in any case, lets first converge on whether or not we want to generally allow to mutate completions, if so, we will not need a new enum at all.

indexes do I need" (i.e., `spec.completions`) and the concept of "how many pods do I want running
at a time" (i.e., `spec.parallelism`). All indexes are expected to start at the same time, but they
can finish at different times.

The above behavior can be modeled using an Indexed job with
- `spec.completions` equal to `spec.parallelism`
- mutable `spec.completions` and `spec.parallelsim` as long as they are mutated together.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- Allow mutating `spec.completions` for `Indexed` job iff the updated value is equal to `spec.parallelism`.

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

- Change the existing behavior of `NonIndexed` mode.

> - Change the existing behavior of `Indexed` mode for jobs that never mutate.

Reviewer 3: ```suggestion
- Change the existing behavior of `Indexed` mode for jobs that never mutate `.spec.completions` and `.spec.parallelism` 
```
just to make it more explicit for readers

will be restarted.

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.

This workload can be modeled as two `Indexed` Jobs, one for the manager and one for the workers. A 

> headless service is created to set up stable hostnames for the worker pods . The workers job is

Reviewer 3: ```suggestion
headless service is created to set up stable hostnames for the worker pods. The workers job is
```

-->

##### e2e tests

Coverage is achieved via integration tests. But we will add one e2e test:
- Job is allowed to scale down, then up, then successfully finish when all pods in the new range exit with success

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
- <test>: <link to test coverage>
-->

### Graduation Criteria

Since the change is only relaxing validation that doesn't require multiple releases and
requires no changes to the job controller, we will start the feature in beta and enabled
by default since we will not get a lot of value from an alpha release. Just like we 
successfully did for [mutable scheduling directives](https://github.com/kubernetes/enhancements/issues/2926).

Note that if tests reveals a required change that invalidates the above understanding, then we will
revert and start in Alpha.

#### Beta

> - Validtion logic in place to allow mutating `spec.completions`

Reviewer 6: ```suggestion
- Validation logic in place to allow mutating `spec.completions`
```

Reviewer 6: What sort of validation logic?

Reviewer 3: Make it explicit:
```suggestion
Validation logic in place to allow mutating `spec.completions` in tandem with `.spec.parallelism`. 
```

Author: updated

-->

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.

> -->

Reviewer 6: Can this feature be safely rolled back? It seems so in the discussion above, but that should be preserved here for posterity.

Author: added a sentence to that effect

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ElasticIndexedJob
  - Components depending on the feature gate: kube-apiserver
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).

###### Does enabling the feature change any default behavior?

No.

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes. If disabled, kube-apiserver will reject mutating requests to `spec.completions` for `Indexed` jobs. 

Reviewer 6: Can you clarify what happens if a mutated request happens and then the feature has been rolled back?

Author: clarified

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

###### What happens if we reenable the feature if it was previously rolled back?

The api-server will accept mutation requests to `spec.completions` for `Indexed` job. 

###### Are there any tests for feature enablement/disablement?

> No, unit and integration tests will be added.

Reviewer 4: Enablement/disablement is different then testing a feature - it's basically switching on/off the feature in the middle of the test.

Given that your feature boils down to relaxing validation, this is purely in memory, which means that such test isn't really needed.

Author: sg, changed to N/A

Reviewer 6: You can still test the relaxation of validation. Just add an e2e test that tests that we can mutate that field for the beta version.

Reviewer 4: That would be a feature test, not enablement/disablement test really.

feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

The change is opt-in, it doesn't impact already running workloads.

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

N/A

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> It will be tested manually prior to beta launch.

Reviewer 4: Please ensure that this will happen before switching feature gate to beta and update the KEP with test description and results before actual promotion.

Author: will do.

No.

<!--
Things to keep in mind include: additional in-memory state, additional
non-trivial computations, excessive access to disks (including increased log
volume), significant amount of data sent and/or received over network, etc.
This through this both in small and large cases, again with respect to the
[supported limits].

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

Create requests will be rejected.

###### What are other known failure modes?

In a multi-master setup, when the cluster has skewed apiservers, some create requests

> may get accepted and some may get rejected.

Reviewer 3: This requires Detection, Mitigations, Diagnostics and Testing filled in, see the template below.

Author: updated

## Design Details

The controller already counts the number of completed indexes within the range defined by `spec.completions` and
uses that to evaluate if the job completed successfully and to set the `status.Succeeded` count. This is not impacted
if `spec.completions` changes over time. Morever, on scale down, the controller also automatically removes any indexes
outside the range. 

As mentioned above, pod failures may have happened to indexes that are outside the range after `spec.completions`
is updated. Those failures will still count against the backoffLimit of the job and the field `status.Failed` will
not be decremented.

The only needed change is to modify Job update validation in the kube-apiserver to allow updating an `Indexed` Job's

> `spec.completions` if `spec.completions=spec.parallelism` in bot the old and new instance of the Job object.

Reviewer 5: both

Author: fixed
